File: Camera/src/main/java/com/example/camera/preview/MySurfacePreview.java
package com.example.camera.preview;

import android.content.Context;
import android.graphics.PixelFormat;
import android.hardware.Camera;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import com.example.camera.util.CameraInstance;

/**
 * Created by renlei
 * DATE: 15-11-5
 * Time: 下午4:52
 */
public class MySurfacePreview extends SurfaceView implements SurfaceHolder.Callback {

    private final SurfaceHolder surfaceHolder;
    private Handler mHandler;

    public MySurfacePreview(Context context, AttributeSet attrs) {
        super(context, attrs);
        surfaceHolder = getHolder();
        //translucent半透明 transparent透明
        surfaceHolder.setFormat(PixelFormat.TRANSPARENT);
        surfaceHolder.addCallback(this);
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        CameraInstance.get().doOpenCamera(Camera.CameraInfo.CAMERA_FACING_BACK);
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        CameraInstance.get().doStartPreview(surfaceHolder);
        if (mHandler != null) {
            mHandler.postDelayed(() -> mHandler.sendEmptyMessage(CameraInstance.PREVIEW_HAS_STARTED), 1000);
        }
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        CameraInstance.get().doStopPreview();
    }

    public void setHandler(Handler handler) {
        this.mHandler = handler;
    }
}

File: FaceVerification/src/main/java/com/example/camera/preview/MySurfacePreview.kt
package com.example.camera.preview

import android.content.Context
import android.graphics.PixelFormat
import android.hardware.Camera
import android.os.Handler
import android.util.AttributeSet
import android.view.SurfaceHolder
import android.view.SurfaceView
import com.example.camera.util.CameraUtil

/**
 * Created by renlei
 * DATE: 15-11-5
 * Time: 下午4:52
 */
// : 表示 extends , 表示 implements
class MySurfacePreview(context: Context, attrs: AttributeSet) : SurfaceView(context, attrs), SurfaceHolder.Callback {

    // val 表示 final
    private val mSurfaceHolder: SurfaceHolder = holder
    // ? = 对象初始化
    private var mHandler: Handler? = null

    // 在构造方法中所做的初始化工作
    init {
        // translucent半透明 transparent透明
        mSurfaceHolder.setFormat(PixelFormat.TRANSPARENT)
        mSurfaceHolder.addCallback(this)
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        // 开启相机
        CameraUtil.instance.doOpenCamera(Camera.CameraInfo.CAMERA_FACING_BACK)
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        // 开始预览
        CameraUtil.instance.doStartPreview(mSurfaceHolder)
        if (mHandler != null) {
            // !! 表示非空对象，可以直接调用属性或者方法
            mHandler!!.postDelayed({ mHandler!!.sendEmptyMessage(CameraUtil.PREVIEW_HAS_STARTED) }, 1000)
        }
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        // 停止预览
        CameraUtil.instance.doStopPreview()
    }

    fun setHandler(handler: Handler) {
        mHandler = handler
    }
}
